# Machine Learning and Statistics - Tasks


## Task 1 - Square Root

## Introduction 

This assignment demonstrates how to create a Python function called  $sqrt2$  that calculates and prints to the screen the square root of 2 to 100 decimal places. This code cannot depend on any of the modules available in the standard library. 


## Research 

"The early Pythagoreans were convinced that every conceivable number could in principle be written in fractional form, as the ratio of two natural numbers. Since there is an infinite supply of natural numbers, they thought, there must be enough to do the job. The discovery that this was an erroneous belief, possibly by the geometer Hippasus in the 5th century BCE, was shocking news. According to legend, Hippasus was hurled off a boat and drowned to prevent the truth becoming widely known, such was its threat to the Pythagorean concept of order in the Universe!." [1] 

Numbers that cannot be expressed as a ratio of natural numbers are called irrational numbers, even though they make perfect sense to modern mathematicians. It is actually easy to see why some numbers are irrational. A famous example is the square root of 2, which is roughly 1.4142, and denoted √2. 

Before one can attempt to write Python code for ...sqrt2..., it is important to gain a better understanding of an effecient method of calculation. "Newton’s method also known as Newton-Raphson method is a root-finding algorithm that produces successively better approximations of the roots of a real-valued function." [2]

                $$ x_{n+1}=\dfrac{1}{2}\left( x_{n}+\dfrac{a}{x_{n}}\right) $$


The approximations of the root go as:

$$ x_(n+1) = x_n - f(x_n) / f’(x_n) $$
$$ x_0 $$  is the rough approximation of the root done at the first and the successive approximations go as  $$ x_1, x_2,...$$ 
$$ f(x_n) $$ is the function whose root is to be determined and $$ f’(x_n) $$ is the derivative of the function.

"Newton's Method is used to find successive approximations to the roots of a function. If the function is y = f(x) and x0 is close to a root, then we usually expect the formula below to give x1 as a better approximation. Then you plug the x1 back in as x0 and iterate" (2)

Newton's Method for square root goes as follows;(4) If we have to find the square root of a number n, the function would be $$ f(x) = x² - N $$ and we would have to find the root of the function,  $$ f(x) $$

Here, the value $$ f(x_n) at x = x_n $$ is $$ f(x_n) = x_n² - N $$

And, the derivative at the point is $$ f’(x_n) = 2 * x_n $$

Now, the better approximation can be found using [1]

$$ x_(n+1) = x_n - (x_n² - N) / (2 * x_n) $$
$$ x_(n+1) = x_n - x_n² / (2 * x_n) + N/ (2 * x_n) $$
$$ x_(n+1) = x_n - x_n / 2+ N/ (2 * x_n) $$
$$ x_(n+1) = x_n / 2+ N/ (2 * x_n) $$
$$ x_(n+1) = (x_n + N/ x_n) / 2 $$


Integer square root of a number is the floor of the square root. The algorithm can be modified a little to find integer square root of a number. The while condition here would be approximate * approximate > N. The algorithm terminates when the approximate squared is less than or equal to N.
The iteration relation here is:

$$ x_(n+1) = (x_n + N // x_n) // 2 $$ where // is integer division [4]

A classic analysis text (Rudin, Principles of Mathematical Analysis) approaches the proof of convergence of this algorithm as follows: we prove that the sequence converges monotonically and is bounded, and hence it has a limit; we then easily see that the limit is √ [5]

This method can be applied to the calculating the square root of 2. If we start with x_1=1 

The number of accurate digits approximately doubles on each iteration. This is a very efficient concergence rate.

![image](https://raw.githubusercontent.com/NiamhOL/Machine-Learning-and-Statistics-2020-Assignments/main/images/Newton.PNG) 

### AdvantagesDisadvantages of Newton's Method

When it converges, Newton's method usually converges very quickly and this is its main advantage. However, Newton's method is not guaranteed to converge and this is obviously a big disadvantage especially compared to the bisection and secant methods which are guaranteed to converge to a solution (provided they start with an interval containing a root).

Newton's method also requires computing values of the derivative of the function in question. This is potentially a disadvantage if the derivative is difficult to compute. 

The stopping criteria for Newton's method differs from the bisection and secant methods. In those methods, we know how close we are to a solution because we are computing intervals which contain a solution. In Newton's method, we don't know how close we are to a solution. All we can compute is the value  and so we implement a stopping criteria based on."[3]



## Square Root 2 in Python 

Loops are often used in programs that compute numerical results by starting with an approximate answer and iteratively improving it. Suppose that you want to know the square root of n. If you start with almost any approximation, you can compute a better approximation with the following formula:


The following implementation of Newton’s method requires two parameters. The first is the value whose square root will be approximated. The second is the number of times to iterate the calculation yielding a better result [7]

In [8]:
def newton_method(number, number_iters = 500):    # https://medium.com/@sddkal/newton-square-root-method-in-python-270853e9185d
    a = float(number) # number to get square root of
    for i in range(number_iters): # iteration number
        number = 0.5 * (number + a / number) # update
 # x_(n+1) = 0.5 * (x_n +a / x_n)
    return number

print (newton_method(2))

1.414213562373095


In [3]:
x = 2 * 10 ** 200  
r = x

def sqrt2(x, r):
    d0 = abs(x - r**2)
    dm = abs(x - (r-1)**2)
    dp = abs(x - (r+1)**2)
    minimised = d0 <= dm and d0 <= dp
    below_min = dp < dm
    return minimised, below_min

while True:
    oldr = r
    r = (r + x // r) // 2

    minimised, below_min = sqrt2(x, r)
    if minimised:
        break

    if r == oldr:
        if below_min:
            r += 1
        else:
            r -= 1
        minimised, _ = sqrt2(x, r)
        if minimised:
            break

print(f'{r // 10**100}.{r % 10**100:0100d}') # [8]

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


"Newton’s algorithm will eventually reach a point where the new approximation is no better than the previous. At that point, we could simply stop. In other words, by repeatedly applying this formula until the better approximation gets close enough to the previous one, we can write a function for computing the square root that uses the number of iterations necessary and no more.

This implementation, shown in codelens, uses a while condition to execute until the approximation is no longer changing. Each time through the loop we compute a “better” approximation using the formula described earlier. As long as the “better” is different, we try again. Step through the program and watch the approximations get closer and closer." [7]

## Validating the above method

In [7]:
from decimal import *
getcontext().prec = 101 # Can try other numbers here
Decimal(2).sqrt()

Decimal('1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727')

In [ ]:
# https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil [10]
sqrt2byNasa = "1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727"
# check I have 100 decimal places places/ 101 significant places.
import re
# verify 100 decimal places
print("number of significant digits",len(str(sqrt2byNasa)[1:]))


## Conclusions 

The algorithm for Newton’s method for numerically approximating a root of a function can be summarised as follows:

* Given a function $$ f(x) $$ of a variable x and a way to compute both $$ f(xi) $$ and its derivative $$ f′(xi)$$ at a given point $$ xi.$$
* Set values for the desired numerical relative precision Δd of the result as well as for the maximum number of iterations N the algorithm is allowed to take.
* Choose an initial guess $$ x0 $$, ideally close to the location of the root to be approximated.
* Compute the next approximation x1 for the root as
$$ x1=x0–f(x0)f′(x0) $$
* Repeat as $$ xi+1=xi–f(xi)f′(xi)$$ and compute the relative difference between the approximations as
$$ Δ=|xi+1−xi||xi|=|f(xi)||f′(xi)xi| $$ until one of two things happens:
1. the maximum number of allowed iterations is reached, i.e., $$ i>N $$
2. the desired numerical relative precision is reached, i.e., $$ Δ<Δd $$

If the algorithm reaches the maximum number of iterations that you set, this can but does not have to indicate that convergence is not possible. Experimentation with suitable values for both Δd and N may be needed to arrive at a satisfactory result. [9] 

### References and Biblography for Task 1

1. https://cosmosmagazine.com/mathematics/the-square-root-of-2/

2. https://hackernoon.com/calculating-the-square-root-of-a-number-using-the-newton-raphson-method-a-how-to-guide-yr4e32zo

3. http://www.cs.utsa.edu/~wagner/CS3343/newton/sqrt.html

4. https://www.math.ubc.ca/~pwalls/math-python/roots-optimization/newton/

5. https://medium.com/@surajregmi/how-to-calculate-the-square-root-of-a-number-newton-raphson-method-f8007714f64

6. https://math.mit.edu/~stevenj/18.335/newton-sqrt.pdf

7. https://runestone.academy/runestone/books/published/thinkcspy/MoreAboutIteration/NewtonsMethod.html

8. https://stackoverflow.com/questions/64278117/is-there-a-way-to-create-more-decimal-points-on-python-without-importing-a-libra
 
9. https://computingskillset.com/solving-equations/the-newton-raphson-method-explained-details-pictures-python-code/

 
 


## Task 2 - Chi Squared

The Task
The Chi-squared test for independence is a statistical hypothesis test like a t-test. It is used to analyse whether two categorical variables are independent. The Wikipedia article gives the table below as an example,stating the Chi-squared value based on it is approximately 24.6. Use scipy.stats to verify this value and calculate the associated p value. You should include a short note with references justifying your analysis in a markdown cell.

### Introduction

"The Chi-Squared test is a statistical hypothesis test that assumes (the null hypothesis) that the observed frequencies for a categorical variable match the expected frequencies for the categorical variable. The test calculates a statistic that has a chi-squared distribution, named for the Greek capital letter Chi (X) pronounced “ki” as in kite." [1] In the exampl used from Wikipedia, a random sample of 650 residents of a city was taken and their occupation was recorded as "white collar" "blue collar" or "no collar". The null hypothesis is that each person's neighbourhood of residence is independent of the person's occupation classification.

![Image](https://raw.githubusercontent.com/NiamhOL/Machine-Learning-and-Statistics-2020-Assignments/main/images/ChiSq.PNG)

$$ \dfrac{\left( observed-\exp ected\right) ^{2}}{\exp ected}= \dfrac{\left( 90-80.54\right) ^{2}}{80.54}\simeq 1.11 $$

In that "cell" of the table we have

$$ 150\times \dfrac{349}{650}\simeq 80.54 $$

The sum of these quantities over all of the cells is the test statistic; in this case, {\displaystyle \approx 24.6}{\displaystyle \approx 24.6}. Under the null hypothesis, this sum has approximately a chi-squared distribution whose number of degrees of freedom are

If the test statistic is improbably large according to that chi-squared distribution, then one rejects the null hypothesis of independence.[2]

The Chi-Squared test does this for a contingency table, first calculating the expected frequencies for the groups, then determining whether the division of the groups, called the observed frequencies, matches the expected frequencies.

The result of the test is a test statistic that has a chi-squared distribution and can be interpreted to reject or fail to reject the assumption or null hypothesis that the observed and expected frequencies are the same. Variables are considered independent if the observed and expected frequencies are similar, that the levels of the variables do not interact, are not dependent. "The chi-square test of independence works by comparing the categorically coded data that you have collected (known as the observed frequencies) with the frequencies that you would expect to get in each cell of a table by chance alone (known as the expected frequencies." [3] 

We can interpret the test statistic in the context of the chi-squared distribution with the requisite number of degress of freedom as follows:

If Statistic >= Critical Value: significant result, reject null hypothesis (H0), dependent.
If Statistic < Critical Value: not significant result, fail to reject null hypothesis (H0), independent.
The degrees of freedom for the chi-squared distribution is calculated based on the size of the contingency table as:

In terms of a p-value and a chosen significance level (alpha), the test can be interpreted as follows:

If p-value <= alpha: significant result, reject null hypothesis (H0), dependent.
If p-value > alpha: not significant result, fail to reject null hypothesis (H0), independent.
For the test to be effective, at least five observations are required in each cell of the contingency table. [1]


### Chi-Squared and Python

In [1]:
# Import module for storing datafram
import pandas as pd # for dataframe work
import scipy.stats as stats # for calculate chi-square value

# Create df with table data. 

data = {'A':[90, 30, 30, 150], 'B':[60, 50, 40, 150], 'C':[104, 51, 45, 200],
        'D':[95, 20, 35, 150], 'Total':[349, 151, 150, 650]}
df = pd.DataFrame(data,index=['White collar', 'Blue collar', 'No Collar', 'Total'])

# Display the Observed results dataframe table
print("Data Table")
df

Data Table


A    B    C    D  Total
White collar   90   60  104   95    349
Blue collar    30   50   51   20    151
No Collar      30   40   45   35    150
Total         150  150  200  150    650

In [2]:
# Create observed results df. Code adapted from [2.3]
df_obs = df.iloc[0:3, 0:4].copy()

# Display the Observed results dataframe table
print("Observed Results Table")
df_obs

Observed Results Table


A   B    C   D
White collar  90  60  104  95
Blue collar   30  50   51  20
No Collar     30  40   45  35

In [3]:
# Copy the data table to form the expected results table dataframe. Code adapted from
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html
df_exp = df.copy()

i, j = 0, 0
# Loop for 3 rows
while i < 3:
    # Loop for 4 columns
    while j < 4:
        df_exp.iloc[i,j] = df_exp.iloc[-1,j]*df_exp.iloc[i,-1]/df_exp.iloc[-1,-1]
        j += 1
    j = 0
    i += 1
    
# Drop total column and total row. Code adapted from
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html
df_exp = df_exp.drop(['Total'], axis=1).drop(['Total'], axis=0)

# Display Expected results dataframe Table
print("Expected Results Table")
df_exp.round(2)

Expected Results Table


A      B       C      D
White collar  80.54  80.54  107.38  80.54
Blue collar   34.85  34.85   46.46  34.85
No Collar     34.62  34.62   46.15  34.62

In [4]:
# Calculate Partial Chi-squared values
df_chi = ((df_obs - df_exp)**2)/df_exp

# Display partial Chi-squared value dataframe Table
print("Partial Chi-squared value Results Table")  
df_chi.round(2)

Partial Chi-squared value Results Table


A     B     C     D
White collar  1.11  5.24  0.11  2.60
Blue collar   0.67  6.59  0.44  6.33
No Collar     0.62  0.84  0.03  0.00

In [5]:
# Calculate full Chi-squared value. Code adapted from
# https://www.geeksforgeeks.org/python-pandas-dataframe-sum
#  https://medium.com/@nhan.tran/the-chi-square-statistic-p3-programming-with-python-87eb079f36af
chi2_man = df_chi.sum().sum()

print(f"The manually calculated chi-square value is ~{chi2_man:.1f}")

The manually calculated chi-square value is ~24.6


In [6]:
# Calculate Chi-square value using scipy.stats library. Code adapted from
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html
chi2_ss, p_ss, dof_ss, exp_ss = stats.chi2_contingency(df_obs)

# Disply the results of the Chi-square test
print(f"""The scipy.stats calculated chi-squared test value is ~{chi2_ss:.1f}
with a p-value of {p_ss:.6f} and degrees of freedom of {dof_ss}.""")

The scipy.stats calculated chi-squared test value is ~24.6
with a p-value of 0.000410 and degrees of freedom of 6.


In [7]:
# Caluclate critical value for a probability level of 5% (0.95). Code adapted from 
#  https://medium.com/@nhan.tran/the-chi-square-statistic-p3-programming-with-python-87eb079f36af
crit_ss = stats.chi2.ppf(q=0.95, df=dof_ss)
print(f"The scipy.stats calculated critical value is {crit_ss:.1f}")

The scipy.stats calculated critical value is 12.6



The Chi-squared value and p-value can be calculated using the scipy.stats.chi2_contingency method and the critical Chi-squared value can be calculated using the stats.chi2.ppf method. The calculate Chi-squared value (24.6) is higher than the critical value (12.6) for a 5% significance level and degrees of freedom in the sampled data. As such, we can reject the null hypotheses that the categories are independent of each other.

### References and Biblography for Task 2

1. https://machinelearningmastery.com/chi-squared-test-for-machine-learning/
2. https://en.wikipedia.org/wiki/Chi-squared_test
3. Urdan, T., 2010. Statistics In Plain English. New York: Routledge.

* https://www.geeksforgeeks.org/different-ways-to-create-pandas-dataframe/ and
* https://stackoverflow.com/a/60909202
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html
* https://www.geeksforgeeks.org/python-pandas-dataframe-sum
* https://medium.com/@nhan.tran/the-chi-square-statistic-p3-programming-with-python-87eb079f36af
* https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html
* https://medium.com/@nhan.tran/the-chi-square-statistic-p3-programming-with-python-87eb079f36af

## Task 3 - Standard Deviation

The standard deviation of an array of numbers x is calculated using numpy as np.sqrt(np.sum((x - np.mean(x))**2)/len(x). However, Microsoft Excel has two different versions of the standard deviatio calculation, STDEV.P and STDEV.S . The STDEV.P function performs the above calculation but in the STDEV.S calculation the division is by len(x)-1 rather
than len(x) . Research these Excel functions, writing a note in a Markdown cell about the difference between them. Then use numpy to perform a simulation demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample. Note that part of this task is to figure out the terminology in the previous sentence.


### Standard Deviation

"The standard deviation is a measure of the spread of scores within a set of data. Usually, we are interested in the standard deviation of a population. However, as we are often presented with data from a sample only, we can estimate the population standard deviation from a sample standard deviation. These two standard deviations - sample and population standard deviations - are calculated differently." [1]  deviation tells you how spread out the data is. It is a measure of how far each observed value is from the mean. In any distribution, about 95% of values will be within 2 standard deviations of the mean.


![image](https://raw.githubusercontent.com/NiamhOL/Machine-Learning-and-Statistics-2020-Tasks/main/images/Std.PNG) 


Generally we are more interested in the population standard deciation as our population contains all the values we are interested in. However, in statistics, we are often only presented with a sample of the population and therefore have to estimate or generalise a population. In this case, were we wish to make a statement about the population standard deviation from which a sample is drawn, we use the sample standard deviation. 

SD =\sqrt{\dfrac{\sum \left| x-\mu \right| ^{2}}{N}}



### Microsoft Excel STDEV.P function

### Microsoft Excel STDEV.S function

In [5]:
# For statistics
import scipy.stats as ss
# For plotting
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = (10.0, 8.0) # Make nice big figures
# For arrays
import numpy as np

### References and Biblography for Task 3

1. https://statistics.laerd.com/statistical-guides/measures-of-spread-standard-deviation.php
